bed/modify.go

```diff
@@ -92,3 +92,178 @@ func MergeHighMem(records []Bed, mergeAdjacent bool) []Bed {
+			answer = append(answer, currAnswer)
+			currAnswer.Chrom = records[i].Chrom
+			currAnswer.ChromStart = midpoint + 1
+			currAnswer.ChromEnd = records[i].ChromEnd
+			currAnswer.Name = records[i].Name
+			currAnswer.Score = records[i].ChromStart
+		}
+	}
+
+	currAnswer.ChromEnd = genome[currAnswer.Chrom].Size
+	answer = append(answer, currAnswer)
+
+	return answer
+}
+
+//helper fucntion of FillSpaceHiddenValue. Define the 3dMidpoint between bed entries A and B (with hidden values
+//Ha and Hb) as (A.ChromEnd-Ha + B.ChromStart+Hb) / 2 s.t. A.ChromEnd < B.ChromStart.
+// if 3dMidpoint(A, B) < A.ChromEnd, then entry A can never have an entry in the output bed of FillSpaceHiddenValue,
+// and should be removed.
+func removeBedsWithNoTerritory(records []Bed) ([]Bed, bool) {
+	var answer []Bed
+	var threeDmidpoint int
+	var violation = false
+	for i := 1; i < len(records); i++ {
+		if records[i-1].Chrom == records[i].Chrom {
+			threeDmidpoint = (records[i-1].ChromEnd - records[i-1].Score + records[i].ChromStart + records[i].Score) / 2
+			if threeDmidpoint < records[i-1].ChromEnd {
+				violation = true
+			} else {
+				answer = append(answer, records[i-1])
+			}
+		} else {
+			answer = append(answer, records[i-1])
+		}
+	}
+	answer = append(answer, records[len(records)-1])
+
+	return answer, violation
+}
+
+//runUntilNoNewHidden directs our mergeKeepLowNameAndScore function to keep updating until no changes occur
+func runUntilNoNewHidden(records []Bed) []Bed {
+	var newHidden bool
+	records, newHidden = mergeKeepLowNameAndScore(records)
+	if newHidden {
+		runUntilNoNewHidden(records)
+	}
+	return records
+}
+
+// FillSpaceHiddenValue accepts a slice of Bed structs and a reference genome as a map[string]chromInfo.ChromInfo and returns a
+// slice of Bed structs that assigns each genomic position to the nearest feature in 3D space from the input bed, using
+// the input bed scores to represent "hidden values", or the distance from that position to its nearest TSS in 3D space.
+func FillSpaceHiddenValue(records []Bed, genome map[string]chromInfo.ChromInfo) []Bed {
+	records = runUntilNoNewHidden(records)
+	records = removeRecordsOnEmptyChrom(records, genome)
+	var violation = true
+	for violation {
+		records, violation = removeBedsWithNoTerritory(records)
+	}
+	var answer = make([]Bed, 0)
+	var threeDMidpoint int
+	var currAnswer = Bed{Chrom: records[0].Chrom, ChromStart: 0, ChromEnd: records[0].ChromEnd, Name: records[0].Name, FieldsInitialized: 4}
+
+	for i := 1; i < len(records); i++ {
+		if records[i].Chrom != currAnswer.Chrom {
+			currAnswer.ChromEnd = genome[records[i-1].Chrom].Size
+			if currAnswer.ChromEnd < currAnswer.ChromStart {
+				log.Print(records[i-1])
+				log.Print(currAnswer)
+				log.Fatalf("Died on new chrom.")
+			}
+			answer = append(answer, currAnswer)
+			currAnswer.Chrom = records[i].Chrom
+			currAnswer.ChromStart = 0
+		} else if currAnswer.Name == records[i].Name && currAnswer.Chrom == records[i].Chrom {
```

Comment:
Is `currAnswer.Chrom == records[i].Chrom` here unnecessary since it's implied in the else if?

Am I right to think that this is the case for "multiple points describing the same gene, possibly TSS + contact point, or contact point + contact point". In this case, expand the gene's bed to be the most inclusive (smallest chromStart, largest chromEnd, lowest score because that entire region can be considered that smallest distance to TSS)? If so, then I think already "remove the contact point / bedpe foot that is closer to the TSS to begin with" (the point Craig mentioned at lab meeting 6/14).